File: src/main/irc/features/sts.ts
/**
 * Strict Transport Security (STS) — IRCv3
 *
 * When connecting over plaintext:
 * - Server advertises sts capability: sts=port=6697,duration=2592000
 * - Client MUST disconnect and reconnect using TLS on the advertised port
 * - Client caches the policy for `duration` seconds
 *
 * When connecting over TLS:
 * - Client should cache the policy
 * - duration=0 means clear the cached policy
 *
 * The STS policy cache is managed by the storage layer.
 */

export interface STSPolicy {
  host: string
  port: number
  duration: number
  cachedAt: string // ISO 8601
}

/** In-memory cache, backed by the database so a policy survives a restart */
const stsPolicies = new Map<string, STSPolicy>()

/** Where policies are kept between runs; injected so this stays testable */
let store: {
  save: (policy: STSPolicy) => void
  forget: (host: string) => void
} | null = null

export function persistSTSPoliciesWith(backing: typeof store): void {
  store = backing
}

/**
 * Check if a host has an active STS policy.
 * Returns the policy if valid, null otherwise.
 */
export function getSTSPolicy(host: string): STSPolicy | null {
  // A hostname, not an IRC name: DNS folds ASCII case and nothing else, so
  // `toLowerCase` is right here and `casemap` would be wrong.
  const policy = stsPolicies.get(host.toLowerCase())
  if (!policy) return null

  // Check if expired
  const cachedAt = new Date(policy.cachedAt).getTime()
  const expiresAt = cachedAt + policy.duration * 1000
  if (Date.now() > expiresAt) {
    stsPolicies.delete(host.toLowerCase())
    store?.forget(host)
    return null
  }

  return policy
}

/**
 * Cache an STS policy (called when server advertises sts capability).
 */
export function setSTSPolicy(host: string, port: number, duration: number): void {
  if (duration === 0) {
    // duration=0 is the server withdrawing the policy
    stsPolicies.delete(host.toLowerCase())
    store?.forget(host)
    return
  }

  const policy: STSPolicy = {
    host: host.toLowerCase(),
    port,
    duration,
    cachedAt: new Date().toISOString()
  }
  stsPolicies.set(policy.host, policy)
  store?.save(policy)
}

/**
 * Where this server must actually be reached.
 *
 * Returns the port and TLS setting to dial with, which is the whole purpose of
 * the policy — a cached STS that nothing consults protects nobody.
 */
export function stsUpgradeFor(
  host: string,
  port: number,
  tls: boolean
): { port: number; tls: true } | null {
  const policy = getSTSPolicy(host)
  if (!policy) return null
  if (tls && port === policy.port) return null
  return { port: policy.port, tls: true }
}

/**
 * Load cached policies from storage (called at startup).
 */
export function loadSTSPolicies(policies: STSPolicy[]): void {
  for (const policy of policies) {
    stsPolicies.set(policy.host.toLowerCase(), policy)
  }
}

/**
 * Get all current policies (for persisting to storage).
 */
export function getAllSTSPolicies(): STSPolicy[] {
  return Array.from(stsPolicies.values())
}

/**
 * Parse the STS capability value.
 * Format: port=6697,duration=2592000
 */
export function parseSTSValue(value: string): { port: number; duration: number } | null {
  const parts: Record<string, string> = {}
  for (const segment of value.split(',')) {
    const eqIdx = segment.indexOf('=')
    if (eqIdx !== -1) {
      parts[segment.slice(0, eqIdx)] = segment.slice(eqIdx + 1)
    }
  }

  const port = parseInt(parts['port'] || '', 10)
  const duration = parseInt(parts['duration'] || '', 10)

  if (isNaN(port) || isNaN(duration)) return null

  return { port, duration }
}

File: android/app/src/main/java/org/switchboard/android/irc/Sts.kt
package org.switchboard.android.irc

import java.time.Instant

/**
 * Strict Transport Security — IRCv3.
 *
 * A server advertising `sts` is telling us it is TLS-only: reconnect on this
 * port, over TLS, and remember that for the stated duration. A client that
 * ignores it keeps offering an attacker a plaintext window — and this is the
 * phone, which is the device most likely to be on somebody else's network.
 *
 * The policy has to outlive the process to be worth anything, so it is written
 * through [Store] rather than kept only in memory.
 */
data class StsPolicy(
    val host: String,
    val port: Int,
    val durationSeconds: Long,
    val cachedAt: Instant
) {
    fun expiredAt(now: Instant): Boolean =
        now.isAfter(cachedAt.plusSeconds(durationSeconds))
}

object Sts {

    /** Where policies live between runs */
    interface Store {
        fun load(): List<StsPolicy>
        fun save(policy: StsPolicy)
        fun forget(host: String)
    }

    private val policies = mutableMapOf<String, StsPolicy>()
    private var store: Store? = null

    @Synchronized
    fun useStore(backing: Store) {
        store = backing
        policies.clear()
        for (policy in backing.load()) policies[policy.host.lowercase()] = policy
    }

    /**
     * Parse the capability's value: `port=6697,duration=2592000`.
     *
     * Anything without both parts is not a policy, and guessing at a partial
     * one would mean redirecting a connection on the strength of a typo.
     */
    fun parse(value: String?): Pair<Int, Long>? {
        if (value.isNullOrBlank()) return null
        val parts = value.split(",").mapNotNull {
            val at = it.indexOf('=')
            if (at < 0) null else it.substring(0, at) to it.substring(at + 1)
        }.toMap()

        val port = parts["port"]?.toIntOrNull() ?: return null
        val duration = parts["duration"]?.toLongOrNull() ?: return null
        if (port !in 1..65535 || duration < 0) return null
        return port to duration
    }

    /** Record what a server has told us, or drop it when duration is zero */
    @Synchronized
    fun remember(host: String, port: Int, durationSeconds: Long) {
        // A hostname, not an IRC name: DNS folds ASCII case and nothing else,
        // so `lowercase` is right here and `casemap` would be wrong.
        val key = host.lowercase()
        if (durationSeconds == 0L) {
            policies.remove(key)
            store?.forget(key)
            return
        }
        val policy = StsPolicy(key, port, durationSeconds, Instant.now())
        policies[key] = policy
        store?.save(policy)
    }

    /** The live policy for a host, forgetting it once it has expired */
    @Synchronized
    fun policyFor(host: String, now: Instant = Instant.now()): StsPolicy? {
        val key = host.lowercase()
        val policy = policies[key] ?: return null
        if (policy.expiredAt(now)) {
            policies.remove(key)
            store?.forget(key)
            return null
        }
        return policy
    }

    /**
     * Where this server must actually be reached.
     *
     * Null when the settings already satisfy the policy, or when there is none.
     */
    @Synchronized
    fun upgradeFor(host: String, port: Int, tls: Boolean): Pair<Int, Boolean>? {
        val policy = policyFor(host) ?: return null
        if (tls && port == policy.port) return null
        return policy.port to true
    }

    /** For tests, and for a device being unpaired */
    @Synchronized
    fun forgetEverything() {
        policies.clear()
        store = null
    }
}
